### DISCUSSION TITLE
Is there a template you've been using to create CLI tools?

### DISCUSSION THREAD

**Discussion by hirehamir** [sharkdp/numbat#415] [General]
`numbat`, `hyperfine` and `bat` are all excellent CLI tools. I was wondering if there's some template you've been using as a foundation for each of them, or if you more or less create each terminal application from scratch each time.

**Comment by hirehamir** [comment]
I've heard positive things about https://github.com/ratatui-org/ratatui, but I don't know if you've ever used it to build anything.

**Comment by eminence** [comment]
To see what numbat was built with, take a look at the [Cargo.toml](https://github.com/sharkdp/numbat/blob/master/numbat-cli/Cargo.toml) file.    The most interesting things here are `rustyline` and `colored`.

Ratatui is great for certain types of TUIs, but it's not used here for numbat.  I personally use ratatui in several of my projects (like [procdump](https://github.com/eminence/procdump) for example) and it's great

**Reply by hirehamir** [reply]
I see. Thanks for the response!

**Comment by sharkdp** [comment]
> numbat, hyperfine and bat are all excellent CLI tools. I was wondering if there's some template you've been using as a foundation for each of them, or if you more or less create each terminal application from scratch each time.

Thank you for the feedback. More or less from scratch every time, yes. But you will certainly find a lot of similarities. Libraries that I use, error handling strategy, etc. If you want to see a smaller example, maybe check out https://github.com/sharkdp/hexyl which is not as big as the other projects.

**Reply by hirehamir** [reply]
You're welcome!
